packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -0,0 +1,156 @@
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public set hashedDocumentId(docId: string | undefined) {
+        assert(this._hashedDocumentId === undefined, "DocId should be set only once");
+        assert(docId !== undefined, "Passed docId should not be undefined");
+        this._hashedDocumentId = docId;
+    }
+
+    public get hashedDocumentId(): string | undefined {
+        return this._hashedDocumentId;
+    }
+
+    public get fluidEpoch() {
+        return this._fluidEpoch;
+    }
+
+    public async fetchFromCache<T>(entry: ICacheEntry, maxOpCount: number | undefined): Promise<T | undefined> {
+        const value = await this.persistedCache.get(entry, maxOpCount);
+        if (value !== undefined) {
+            try {
+                this.validateEpochFromResponse(value.fluidEpoch);
+            } catch (error) {
+                this.checkForEpochError(createOdspNetworkError("Epoch Mismatch", fluidEpochMismatchError));
+            }
+            return value.value as T;
+        }
+    }
+
+    /**
+     * Api to fetch the response for given request and parse it as json.
+     * @param url - url of the request
+     * @param fetchOptions - fetch options for request containing body, headers etc.
+     * @param addInBody - Pass True if caller wants to add epoch in post body.
+     */
+    public async fetchAndParseAsJSON<T>(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<IOdspResponse<T>> {
+        // Add epoch in fetch request.
+        const request = this.addEpochInRequest(url, fetchOptions, addInBody);
+        try {
+            const response = await fetchAndParseAsJSONHelper<T>(request.url, request.fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    /**
+     * Api to fetch the response as it is for given request.
+     * @param url - url of the request
+     * @param fetchOptions - fetch options for request containing body, headers etc.
+     * @param addInBody - Pass True if caller wants to add epoch in post body.
+     */
+    public async fetchResponse(
+        url: string,
+        fetchOptions: {[index: string]: any},
+        addInBody: boolean = false,
+    ): Promise<Response> {
+        // Add epoch in fetch request.
+        const request = this.addEpochInRequest(url, fetchOptions, addInBody);
+        try {
+            const response = await fetchHelper(request.url, request.fetchOptions);
+            this.validateEpochFromResponse(response.headers.get("x-fluid-epoch"));
+            return response;
+        } catch (error) {
+            this.checkForEpochError(error);
+            throw error;
+        }
+    }
+
+    private addEpochInRequest(
+        url: string,
```

Comment:
(together with comment below): Maybe it's worth to have a small class that has base URI,  various arguments (to become query params or headers) and token. And than various code can keep adding it, and at the end there is a method to "pack" results into url & headers.

Reply:
I will do a follow up PR on code refactoring related to this. 
Don't want to include that in this PR as it require some more changes to odspDocumentStorageManager also.